Device App Docs (#7461)
auth0/docs: articles/flows/concepts/device-auth.md

---
title: Device Authorization Flow
description: Learn how the Device Authorization flow works and why you should use it for input-constrained devices, such as smart TVs and media consoles. For use with native apps.
topics:
  - input-constrained-devices
  - client-credentials
  - api-authorization
  - grants
  - flows
  - authorization
  - mobile-apps
  - desktop-apps
  - native-apps
contentType: concept
useCase:
  - add-login
  - secure-api
  - call-api
---
# Device Authorization Flow

> With devices that connect wirelessly to the internet, rather than authenticate the user directly, the device asks the user to go to a link on their computer or smartphone and authorize the device. This avoids a poor user experience for devices that do not have an easy way to enter text. To do this, device apps use the Device Authorization Flow (drafted in [OAuth 2.0](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15)), in which they pass along their Client ID to initiate the authorization process and get a token.

Reviewer: 🤔 It should not matter how the device is connected to internet, it could use ethernet. Can't we just say "With input-constrained devices that connect to the internet, rather than..."?

## How it works

The Device Authorization Flow contains two different paths; one occurs on the device requesting authorization and the other occurs in a browser. The browser flow path, wherein a device code is bound to the session in the browser, occurs in parallel to part of the device flow path.

> ![Device Authorization Sequence](/media/articles/flows/concepts/auth-sequence-device-auth.png)

Reviewer: Use `/oauth/device/code` and `/oauth/token`

Reviewer: Should we mention that the user needs to confirm in step (1) in browser? maybe it's too detailed 🤔 
In steps 4 for browser, since the device path left the authorization server on "Determine whether the user authorized the device", should it show "Authorize device" message? That way the condition whether the device path got stopped gets satisfied.

Reviewer 2: @ziluvatar, thanks for pointing out the problem with the endpoint names. This was the language we used on all the other flow docs (/token) and in some other areas of the site, so I followed suit, but I see that it is inaccurate. I'll fix this here and open a separate PR to take care of the other site areas. Great feedback!

Reviewer 2: @ziluvatar, we saw step 3 of the browser flow (the "consent" piece) as being the authorization, which is how it is represented in our other flows. That said, we did struggle with how to show the two flows visually.

Reviewer 2: I have a request in to graphics to update the image with the proper endpoint names.

Reviewer: > we saw step 3 of the browser flow (the "consent" piece) as being the authorization, which is how it is represented in our other flows. 

Ok... consent is optional (and authenticate too in case of using SSO, but that's another story), in any case in step 4, for me, it is the authorization server (Auth0 Tenant) the one that does something, and it does not generate an access token, it activates or marks as authorized the device.
Maybe it is like that in other flows because Auth0 Tenant would return the tokens to the client in next step, however, with two flows here: either the generation goes to the device flow or we should rename the step for the browser flow.

Reviewer 2: Makes sense. Ok, I've asked Stacy to change step 4 in the Browser Flow to "Marks device as authorized". Will leave this open until I get the graphic back and have added it to the PR.

Reviewer 2: @ziluvatar , diagram has been updated.

### Device Flow

1. The user starts the app on the device.

> 2. The device app requests authorization from the Auth0 Authorization Server using its Client ID (**/device/code** endpoint).

Reviewer: ```suggestion
2. The device app requests authorization from the Auth0 Authorization Server using its Client ID (**/oauth/device/code** endpoint).
```

> 3. The Auth0 Authorization Server responds with a `device_code`, `user_code`, `verification_uri`, `expires_in` (lifetime in seconds for `device_code` and `user_code`), and polling `interval`.

Reviewer: ```suggestion
3. The Auth0 Authorization Server responds with a `device_code`, `user_code`, `verification_uri`, `verification_uri_complete`, `expires_in` (lifetime in seconds for `device_code` and `user_code`), and polling `interval`.
```

> 4.  The device app asks the user to activate using their computer or smartphone and displays the `verification_uri` and `user_code`. (Instead of the user code, the app could also display a QR code or shortened URL with embedded user code.)

Reviewer 3: Explain how `verification_uri_complete` can be used for generating a QR code. Additionally, also explain that for a native app running on your computer, the url can be used to open a browser.

> 5. The device app begins polling your Auth0 Authorization Server for an Access Token (**/token** endpoint) using the time period specified by `interval`. The device app continues polling until either the user completes the browser flow path or the user code expires.

Reviewer: ```suggestion
5. The device app begins polling your Auth0 Authorization Server for an Access Token (**/oauth/token** endpoint) using the time period specified by `interval`. The device app continues polling until either the user completes the browser flow path or the user code expires.
```

> 5. The device app begins polling your Auth0 Authorization Server for an Access Token (**/oauth/token** endpoint) using the time period specified by `interval`. The device app continues polling until either the user completes the browser flow path or the user code expires.

Reviewer 4: based on our own mishaps in client implementation we should note that each interval should start counting after receipt of the last request's response to avoid `slow_down` errors due to network latency

6. When the user successfully completes the browser flow path, your Auth0 Authorization Server responds with an Access Token (and optionally, a Refresh Token). The device app should now forget its `device_code` because it will expire.
7. Your device app can use the Access Token to call an API to access information about the user.
8. The API responds with requested data.

### Browser Flow

> 1. The user visits the `verification_URL` on their computer. If the `user_code` is not embedded in the `verification_URL`, then the user enters the `user_code`.

Reviewer: ```suggestion
1. The user visits the `verification_uri` on their computer. If the `user_code` is not embedded in the `verification_uri`, then the user enters the `user_code`.
```

Reviewer: Should we mention that the user needs to confirm the device that is being activated is displaying the user code?

> 2. Your Auth0 Authorization Server redirects the user to the login and authorization prompt.

Reviewer: ```suggestion
2. Auth0 Authorization Server redirects the user to the login and consent prompt if needed.
```

> 1. The user visits the `verification_uri` on their computer and confirms that the device that is being activated is displaying the `user_code`. If the `user_code` is not embedded in the `verification_uri`, then the user enters the `user_code`.

Reviewer: ```suggestion
1. The user visits the `verification_uri` on their computer, enters the `user_code` and confirms that the device that is being activated is displaying the `user_code`. If the user visits the `verification_uri_complete` by any other mechanism (such as by scanning a QR code), only the device confirmation will be needed.
```

3. The user authenticates using one of the configured login options and may see a consent page asking to authorize the device app.
4. If the `user_code` is valid, when the user consents, your device app is authorized to access the API.

## How to implement it

The easiest way to implement the Device Authorization Flow is to follow our tutorial: [Add Authorization Using the Device Authorization Flow](/flows/guides/device-auth/add-auth-device-auth).

> ## Revoking Device Authorizations

Reviewer: Not sure if this is accurate. We have "Devices" and we have "Authorized Applications", if the user authorized the device app but the device app didn't ask for a Refresh Token then admins will not see anything on "Devices", however, the grant remains on the "Authorized Applications" tab.
Maybe it's more confusing to add that distinction, or we can simply modify the title to "Revoking Device Offline Access", meaning like if the device app wants to access the API later, they will need to run the device flow again.

> 4. If the `user_code` is valid, when the user consents, your device app is authorized to access the API.

Reviewer: ```suggestion
4. Your device app is authorized to access the API.
```
(suggestion based on the updated image, because the user code was validated on step 1)
